drizzle-orm/src/utils.ts

```diff
@@ -194,6 +211,12 @@ export interface DrizzleConfig<TSchema extends Record<string, unknown> = Record<
 	logger?: boolean | Logger;
 	schema?: TSchema;
 }
+export type ValidateShape<T, ValidShape, TResult = T> = T extends ValidShape
```

Comment:
Set operators allow for the column names to be different between the queries, as long as the data types are the same. I.e. this query, for example, is correct:
```sql
select 1 as a union select 2 as b;
```
Therefore, it should be possible to have a query like this:
```ts
db
  .select({
    a: sql`1`
  })
  .union(
    db.select({
      b: sql`2`
    }),
  )
```

I don't think it's possible to validate this on type level, therefore I don't think we can make use of the `ValidateShape` type. Corresponding changes should be also made to the runtime checks. We should also have a test that covers such cases.

Reply:
This will be by far the hardest change as the whole type safety I built was around the idea of maintaining the same column names.
I'd argue we should keep it the way it is currently implemented. I think allowing different column names will introduce confusion because, the column names are only used in the JS/TS side during mapping.
For example, in the current implementation the following is allowed:
```ts
db
  .select({
    a: sql`1`.as("a")
  })
  .union(
    db.select({
      a: sql`2`.as("b")
    }),
  )
```
Which will produce the following sql:
```sql
select 1 as a union select 2 as b;
```
and is valid sql.

Reply:
If we allow different `SelectedFields` keys, what should the user expect the end type to be?
Furthermore, drizzle has historically not cared about the underline *SQL* column names and maps the results to the `SelectedFields` keys, whatever they are. For example:
```ts
await db.select({ someKey: users.name }).from(users)
```
Has always been of type:
```ts
{ someKey: string }[]
```
And not
```ts
{ name: string }[]
```